File: 545-boundary-of-binary-tree/545-boundary-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    bool isLeaf(TreeNode* root){
        
        if(root==NULL) return false;
        
        if(root->left==NULL and root->right==NULL){
            return true;
        }
        return false;
    }
    
    void addLeftBoundary(TreeNode* root,vector<int>& ans){
        TreeNode* curr = root->left;
        
        while(curr){
            if(!isLeaf(curr)) ans.push_back(curr->val);
            
            if(curr->left) curr = curr->left;
            else curr = curr->right;
        }
    }
    
    
    void addLeaves(TreeNode* root,vector<int>& ans){
        
        if(isLeaf(root)){
            ans.push_back(root->val);
            return;
        }
        
        if(root->left) addLeaves(root->left,ans);
        if(root->right) addLeaves(root->right,ans);
        
        
    }
    
    void addRightBoundary(TreeNode* root,vector<int>& ans){
        
        TreeNode* curr = root->right;
        vector<int> temp;
        
        while(curr){
            if(!isLeaf(curr)) temp.push_back(curr->val);
            
            if(curr->right) curr=curr->right;
            else curr = curr->left;
        }
        
        for(int i=temp.size()-1;i>=0;i--){
            ans.push_back(temp[i]);
        }
        
        
    }
    
    vector<int> boundaryOfBinaryTree(TreeNode* root) {
        
        vector<int> ans;
        
        if(!root) return ans;
        
        if(!isLeaf(root)) ans.push_back({root->val}); // 1st add root to ans;
        
        addLeftBoundary(root,ans);
        addLeaves(root,ans);
        addRightBoundary(root,ans);
        
        return ans;
        
        
    }
};

File: 545-boundary-of-binary-tree/545-boundary-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    
    def addleaf(self,root,ans):
        
        if(self.isLeaf(root)):
            ans.append(root.val)
            return;
        
        if(root.left):
            self.addleaf(root.left,ans)
        
        if(root.right):
            self.addleaf(root.right,ans)
        
        
    
    def isLeaf(self,root):
        if root is None:
            return False
        
        if root.left is None and root.right is None:
            return True
        
        return False
    
    def addleft(self,root,ans):
        
        curr = root.left
        
        while(curr):
            
            if(self.isLeaf(curr)==False):
                ans.append(curr.val)
            
            if curr.left:
                curr = curr.left
            else:
                curr = curr.right
        
    def addright(self,root,ans):
        
        temp = []
        
        curr = root.right
        
        while(curr):
            if(self.isLeaf(curr)==False):
                temp.append(curr.val)
            
            if(curr.right):
                curr = curr.right
            
            else:
                curr = curr.left
        # print(temp)
        ans.extend(temp[::-1])
        
    
    def boundaryOfBinaryTree(self, root: Optional[TreeNode]) -> List[int]:
        
        ans = []
        
        if(root is None):
            return ans
    
        if(not self.isLeaf(root)):
            ans.append(root.val)
            
        self.addleft(root,ans)
        # print(ans)
        self.addleaf(root,ans)
        # print(ans)
        self.addright(root,ans)
        # print(ans)
        return ans
        
        